include/swift/SILOptimizer/Utils/InstOptUtils.h

```diff
@@ -375,7 +387,42 @@ bool tryCheckedCastBrJumpThreading(
 /// cleaning up any dead code resulting from deleting those instructions. Use
 /// this utility instead of
 /// \c recursivelyDeleteTriviallyDeadInstruction.
+///
+/// This is designed to be used with a single 'onDelete' callback, which is
+/// invoked consistently just before deleting each instruction. It's usually
+/// used to avoid iterator invalidation (see the updatingIterator() factory
+/// method). The other InstModCallbacks should generally be handled at a higher
+/// level, and avoided altogether if possible. The following two are supported
+/// for flexibility:
+///
+/// callbacks.createdNewInst() is invoked incrementally when it fixes lifetimes
+/// while deleting a set of instructions, but the SIL may still be invalid
+/// relative to the new instruction.
+///
+/// callbacks.notifyWillBeDeletedFunc() is invoked when a dead instruction is
+/// first recognized and was not already passed in by the client. During the
+/// callback, the to-be-deleted instruction has valid SIL. It's operands and
+/// uses can be inspected and cached. It will be deleted later during
+/// cleanupDeadInstructions().
+///
+/// Note that the forceDelete* APIs only invoke notifyWillBeDeletedFunc() when
+/// an operand's definition will become dead after force-deleting the specified
+/// instruction. Some clients force-delete related instructions one at a
+/// time. They may not force-deleted. It's the client's responsiblity to invoke
+/// notifyWillBeDeletedFunc() on those explicitly deleted instructions if
+/// needed.
+///
 class InstructionDeleter {
+public:
+  static InstructionDeleter updatingIterator(SILBasicBlock::iterator &iter) {
```

Comment:
This works exactly like it does today with the onDelete callback. You need to advance the iterator before potentially deleting the current instruction.

I agree it's easy to get wrong. I'm tempted to provide another abstraction, but I think that should be reviewed in a new PR. It's not straightforward. You'll never be able to use a range-based for loop safely with a "non-deleted instruction iterator" because that requires access to the current instruction, which may have been deleted. You could provide a `next`-style iterator, and I do think that's worth doing.

Reply:
For now, I'll at least comment `updatingIterator` and add a TODO to design an iterator abstraction.

Reply:
@eeckstein Ok, I lied about never using a "range-base" for loop. You can do it as long as you're willing to return nullptr for the current instruction if it was already deleted. I actually think that could be very useful, so went ahead and implemented it:

https://github.com/apple/swift/pull/37669/commits/7b9d8af4bb1df2495f4fc6886ddae5c11f3af420#diff-31e14eaf82b08085838ced1032b31bf8ba537186058636027fb47bbe7cc98f99R2855

Now you can do totally safe iteration:

https://github.com/apple/swift/pull/37669/files#diff-34cd3ed8b9f60f7d19c7762a465e1f1b9eda27644c4bfaae72d65e744d567fbaR480

And it handles new instructions properly, which used to be broken.